File: sensible/py/src/sensible/strings.py
import re
import unicodedata
import random
import string
from typing import Optional, List

def slugify(text: str, separator: str = "-") -> str:
    """Convert text to a URL-friendly slug."""
    text = unicodedata.normalize("NFKD", text).encode("ascii", "ignore").decode("ascii")
    text = re.sub(r"[^\w\s-]", "", text.lower())
    return re.sub(r"[-\s]+", separator, text).strip(separator)

def truncate(text: str, length: int, suffix: str = "...") -> str:
    """Truncate text to a specific length with an optional suffix."""
    if len(text) <= length:
        return text
    return text[:length].rstrip() + suffix

def to_snake(text: str) -> str:
    """Convert text to snake_case."""
    s1 = re.sub("(.)([A-Z][a-z]+)", r"\1_\2", text)
    s2 = re.sub("([a-z0-9])([A-Z])", r"\1_\2", s1).lower()
    return re.sub(r"[-_\s]+", "_", s2).strip("_")

def to_camel(text: str) -> str:
    """Convert text to camelCase."""
    s = to_pascal(text)
    return s[0].lower() + s[1:] if s else ""

def to_pascal(text: str) -> str:
    """Convert text to PascalCase."""
    words = re.split(r"[-_\s]+", text)
    return "".join(word.capitalize() for word in words if word)

def to_kebab(text: str) -> str:
    """Convert text to kebab-case."""
    return to_snake(text).replace("_", "-")

def capitalize(text: str) -> str:
    """Capitalize the first letter of the text."""
    if not text:
        return text
    return text[0].upper() + text[1:]

def lower(text: str) -> str:
    """Convert text to lowercase."""
    return text.lower()

def upper(text: str) -> str:
    """Convert text to uppercase."""
    return text.upper()

def swap_case(text: str) -> str:
    """Swap the case of all letters in the text."""
    return text.swapcase()

def reverse(text: str) -> str:
    """Reverse the text."""
    return text[::-1]

def repeat(text: str, times: int) -> str:
    """Repeat the text a given number of times."""
    return text * times

def pad_left(text: str, length: int, char: str = " ") -> str:
    """Pad text on the left to reach a certain length."""
    if len(text) >= length:
        return text
    return char * (length - len(text)) + text

def pad_right(text: str, length: int, char: str = " ") -> str:
    """Pad text on the right to reach a certain length."""
    if len(text) >= length:
        return text
    return text + char * (length - len(text))

def pad(text: str, length: int, char: str = " ") -> str:
    """Pad text on both sides to reach a certain length (center padding)."""
    if len(text) >= length:
        return text
    padding = length - len(text)
    left = padding // 2
    right = padding - left
    return char * left + text + char * right

def strip_prefix(text: str, prefix: str) -> str:
    """Remove a prefix from the text if it exists."""
    if text.startswith(prefix):
        return text[len(prefix):]
    return text

def strip_suffix(text: str, suffix: str) -> str:
    """Remove a suffix from the text if it exists."""
    if text.endswith(suffix) and suffix:
        return text[:len(text)-len(suffix)]
    return text

def pluralize(word: str, count: int = 2, custom: Optional[str] = None) -> str:
    """Pluralize a word based on count, with optional custom plural form."""
    if count == 1:
        return word
    if custom is not None:
        return custom
    
    # Uncountable nouns (same singular and plural)
    uncountable = {"sheep", "fish", "deer", "species", "aircraft", "salmon", "tuna"}
    if word.lower() in uncountable:
        return word
    
    # Basic English pluralization rules
    if word.endswith("y") and not word.endswith("ay") and not word.endswith("ey") and not word.endswith("oy") and not word.endswith("uy"):
        return word[:-1] + "ies"
    if word.endswith(("s", "sh", "ch", "x", "z")):
        return word + "es"
    return word + "s"

def singularize(word: str) -> str:
    """Singularize a plural word."""
    # Uncountable nouns
    uncountable = {"sheep", "fish", "deer", "species", "aircraft", "salmon", "tuna"}
    if word.lower() in uncountable:
        return word
    
    if word.endswith("ies"):
        return word[:-3] + "y"
    if word.endswith("es") and word.endswith(("ses", "shes", "ches", "xes", "zes")):
        return word[:-2]
    if word.endswith("s"):
        return word[:-1]
    return word

def starts_with(text: str, prefix: str) -> bool:
    """Check if text starts with a prefix."""
    return text.startswith(prefix)

def ends_with(text: str, suffix: str) -> bool:
    """Check if text ends with a suffix."""
    return text.endswith(suffix)

def contains(text: str, substring: str) -> bool:
    """Check if text contains a substring."""
    return substring in text

def count_occurrences(text: str, substring: str) -> int:
    """Count occurrences of a substring in text."""
    return text.count(substring)

def random_string(length: int = 8, chars: str = string.ascii_letters + string.digits) -> str:
    """Generate a random string of a given length."""
    return "".join(random.choice(chars) for _ in range(length))

def is_palindrome(text: str) -> bool:
    """Check if text is a palindrome (ignoring case and spaces)."""
    cleaned = re.sub(r"\W+", "", text.lower())
    return cleaned == cleaned[::-1]

def mask_text(text: str, visible_start: int = 0, visible_end: int = 0, mask_char: str = "*") -> str:
    """Mask text, showing only the specified number of characters at start and end."""
    if len(text) <= visible_start + visible_end:
        return text
    masked_len = len(text) - visible_start - visible_end
    return text[:visible_start] + (mask_char * masked_len) + text[-visible_end:] if visible_end > 0 else text[:visible_start] + (mask_char * masked_len)

def word_count(text: str) -> int:
    """Count the number of words in text."""
    return len(text.split())

def sentence_count(text: str) -> int:
    """Count the number of sentences in text."""
    return len(re.findall(r"[^.!?]+[.!?]+", text))

def first_word(text: str) -> Optional[str]:
    """Get the first word from text."""
    words = text.split()
    return words[0] if words else None

def last_word(text: str) -> Optional[str]:
    """Get the last word from text."""
    words = text.split()
    return words[-1] if words else None

File: sensible/go/strings/strings.go
package strings

import (
	"math/rand"
	"regexp"
	"strings"
	"unicode"

	"golang.org/x/text/cases"
	"golang.org/x/text/language"
)

var (
	reNonAlphaNum = regexp.MustCompile(`[^\w\s-]`)
	reSpaces      = regexp.MustCompile(`[-\s]+`)
	reUpperLower  = regexp.MustCompile(`([a-z0-9])([A-Z])`)
	reSplit       = regexp.MustCompile(`[-_\s]+`)

	// titler is the shared Unicode-correct title caser (language-neutral).
	titler = cases.Title(language.Und, cases.NoLower)
)

// capitaliseWord upper-cases the first rune of a word, lower-cases the rest.
func capitaliseWord(w string) string {
	runes := []rune(w)
	if len(runes) == 0 {
		return ""
	}
	return string(unicode.ToUpper(runes[0])) + strings.ToLower(string(runes[1:]))
}

// Slugify converts text to a URL-friendly slug.
func Slugify(text string, separator string) string {
	result := strings.ToLower(text)
	result = reNonAlphaNum.ReplaceAllString(result, "")
	result = reSpaces.ReplaceAllString(result, separator)
	return strings.Trim(result, separator)
}

// Truncate truncates text to a specific length with an optional suffix.
func Truncate(text string, length int, suffix string) string {
	if len(text) <= length {
		return text
	}
	return strings.TrimSpace(text[:length]) + suffix
}

// ToSnake converts text to snake_case.
func ToSnake(text string) string {
	s := reUpperLower.ReplaceAllString(text, "${1}_${2}")
	s = strings.ToLower(s)
	s = reSplit.ReplaceAllString(s, "_")
	return strings.Trim(s, "_")
}

// ToPascal converts text to PascalCase using Unicode-correct capitalisation.
func ToPascal(text string) string {
	words := reSplit.Split(text, -1)
	var result string
	for _, word := range words {
		if word != "" {
			result += capitaliseWord(word)
		}
	}
	return result
}

// ToCamel converts text to camelCase.
func ToCamel(text string) string {
	s := ToPascal(text)
	if s == "" {
		return ""
	}
	runes := []rune(s)
	return string(unicode.ToLower(runes[0])) + string(runes[1:])
}

// ToKebab converts text to kebab-case.
func ToKebab(text string) string {
	return strings.ReplaceAll(ToSnake(text), "_", "-")
}

// ToTitle converts text to Title Case using Unicode-correct rules.
func ToTitle(text string) string {
	return titler.String(strings.ToLower(text))
}

// Capitalize upper-cases the first Unicode letter of text, preserving the rest.
func Capitalize(text string) string {
	runes := []rune(text)
	if len(runes) == 0 {
		return ""
	}
	return string(unicode.ToUpper(runes[0])) + string(runes[1:])
}

// Reverse reverses text, correctly handling multi-byte Unicode characters.
func Reverse(text string) string {
	runes := []rune(text)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

// Repeat repeats text a given number of times.
func Repeat(text string, times int) string {
	return strings.Repeat(text, times)
}

// PadLeft pads text on the left to reach a certain length.
func PadLeft(text string, length int, char string) string {
	if len(text) >= length {
		return text
	}
	return strings.Repeat(char, length-len(text)) + text
}

// PadRight pads text on the right to reach a certain length.
func PadRight(text string, length int, char string) string {
	if len(text) >= length {
		return text
	}
	return text + strings.Repeat(char, length-len(text))
}

// Pad pads text on both sides to reach a certain length.
func Pad(text string, length int, char string) string {
	if len(text) >= length {
		return text
	}
	padding := length - len(text)
	left := padding / 2
	right := padding - left
	return strings.Repeat(char, left) + text + strings.Repeat(char, right)
}

// StripPrefix removes a prefix from text if it exists.
func StripPrefix(text string, prefix string) string {
	return strings.TrimPrefix(text, prefix)
}

// StripSuffix removes a suffix from text if it exists.
func StripSuffix(text string, suffix string) string {
	return strings.TrimSuffix(text, suffix)
}

// WordCount counts the number of words in text.
func WordCount(text string) int {
	return len(strings.Fields(text))
}

// SentenceCount counts the number of sentences in text.
func SentenceCount(text string) int {
	return len(regexp.MustCompile(`[^.!?]+[.!?]+`).FindAllString(text, -1))
}

// Lower converts text to lowercase.
func Lower(text string) string {
	return strings.ToLower(text)
}

// Upper converts text to uppercase.
func Upper(text string) string {
	return strings.ToUpper(text)
}

// SwapCase swaps the case of all letters in the text.
func SwapCase(text string) string {
	return strings.Map(func(r rune) rune {
		if unicode.IsUpper(r) {
			return unicode.ToLower(r)
		}
		if unicode.IsLower(r) {
			return unicode.ToUpper(r)
		}
		return r
	}, text)
}

// StartsWith checks if text starts with a prefix.
func StartsWith(text string, prefix string) bool {
	return strings.HasPrefix(text, prefix)
}

// EndsWith checks if text ends with a suffix.
func EndsWith(text string, suffix string) bool {
	return strings.HasSuffix(text, suffix)
}

// Contains checks if text contains a substring.
func Contains(text string, substring string) bool {
	return strings.Contains(text, substring)
}

// CountOccurrences counts occurrences of a substring in text.
func CountOccurrences(text string, substring string) int {
	return strings.Count(text, substring)
}

// RandomString generates a random string of a given length.
func RandomString(length int, charset string) string {
	if charset == "" {
		charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
	}
	b := make([]byte, length)
	for i := range b {
		b[i] = charset[rand.Intn(len(charset))]
	}
	return string(b)
}

// IsPalindrome checks if text is a palindrome (ignoring case and non-alphanumeric characters).
func IsPalindrome(text string) bool {
	var cleaned []rune
	for _, r := range strings.ToLower(text) {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			cleaned = append(cleaned, r)
		}
	}
	for i := 0; i < len(cleaned)/2; i++ {
		if cleaned[i] != cleaned[len(cleaned)-1-i] {
			return false
		}
	}
	return true
}

// MaskText masks text, showing only the specified number of characters at start and end.
func MaskText(text string, visibleStart int, visibleEnd int, maskChar string) string {
	runes := []rune(text)
	if len(runes) <= visibleStart+visibleEnd {
		return text
	}
	maskedLen := len(runes) - visibleStart - visibleEnd
	result := string(runes[:visibleStart])
	result += strings.Repeat(maskChar, maskedLen)
	if visibleEnd > 0 {
		result += string(runes[len(runes)-visibleEnd:])
	}
	return result
}

// FirstWord gets the first word from text.
func FirstWord(text string) string {
	fields := strings.Fields(text)
	if len(fields) == 0 {
		return ""
	}
	return fields[0]
}

// LastWord gets the last word from text.
func LastWord(text string) string {
	fields := strings.Fields(text)
	if len(fields) == 0 {
		return ""
	}
	return fields[len(fields)-1]
}
